src/advanced-concepts/system-apis/message-inspection.md

```diff
@@ -1 +1,132 @@
+
+Additionally, the IC provides _message inspection_ functionality to inspect [_update_](/internet-computer-programming-concepts/actors.html#public-shared-update) or [_oneway_](/internet-computer-programming-concepts/actors.html#public-shared-oneway) calls and either _accept_ or _decline_ the call _before_ running a public shared function.
+
+> **NOTE**  
+> _Message inspection is executed by a [single replica](#message-inspection-vs-caller-identifying-functions) and does not provide the full security guarantees of going through consensus._
+
+## The `inspect` system function
+
+In Motoko, this functionality can be implemented as a _system function_ called `inspect`. This function takes an [object](/common-programming-concepts/objects-and-classes/objects.html) as an argument and returns a `Bool` value. If we _name_ the object argument `args`, then the _function signature_ looks like this:
+
+```motoko
+system func inspect(args) : Bool
+```
+
+Note that the [return type](/internet-computer-programming-concepts/actors.html#shared-types) is NOT `async`. Also, this function CAN NOT update the [state of the actor](/internet-computer-programming-concepts/basic-memory-persistence.html).
+
+### How it works
+
+The `inspect` function is run before any call to an [_update_](/internet-computer-programming-concepts/actors.html#public-shared-update) or [_oneway_](/internet-computer-programming-concepts/actors.html#public-shared-oneway) function of an [actor](/internet-computer-programming-concepts/actors.html). The call to the [actor](/internet-computer-programming-concepts/actors.html) is then _inspected_ by the `inspect` system function.
+
+The argument to the `inspect` function (an object we call `args` in the function signature above) is provided by the IC and contains information about:
+
+- The [public shared function](/internet-computer-programming-concepts/actors.html#public-shared-functions-in-actors) that is being called
+- The [caller](/internet-computer-programming-concepts/principals-and-authentication.html#caller-authenticating-public-shared-functions) of the function
+- The _binary content_ of the _message argument_.
+
+The `inspect` function may examine this information about the call and decide to either _accept_ or _decline_ the call by returning either `true` or `false` respectively.
+
+> **NOTE**  
+> _Calls to either [query functions](/internet-computer-programming-concepts/actors.html#public-shared-query), [other canisters](/advanced-concepts/async-programming/cross-canister-calls-and-rollbacks.html) or the [management canister](/common-internet-computer-canisters/ic-management-canister.html) are NOT inspected by the `inspect` system function._
```

Comment:
```suggestion
> _Ingress query calls to [query functions](/internet-computer-programming-concepts/actors.html#public-shared-query) and any calls from [other canisters](/advanced-concepts/async-programming/cross-canister-calls-and-rollbacks.html)  are NOT inspected by the `inspect` system function._
```
Not sure it's worth mentioning the management canister here and you can't author its inspect message anyway.